tempesta_fw/http_sess.c

```diff
@@ -402,51 +438,246 @@ tfw_http_sticky_add(TfwHttpResp *resp)
 	return r;
 }
 
+/*
+ * Calculate HMAC value for redirection mark.
+ *
+ * HMAC value is based on:
+ * - Number of attempts for session establishing;
+ * - Timestamp of first attempt;
+ * - Secret key.
+ */
+static int
+__redir_hmac_calc(TfwHttpReq *req, RedirMarkVal *mv)
+{
+	int r;
+	char desc[sizeof(struct shash_desc)
+		  + crypto_shash_descsize(tfw_sticky_shash)]
+		  CRYPTO_MINALIGN_ATTR;
+	struct shash_desc *shash_desc = (struct shash_desc *)desc;
+
+	bzero_fast(desc, sizeof(desc));
+	shash_desc->tfm = tfw_sticky_shash;
+	shash_desc->flags = 0;
+
+	TFW_DBG("http_sess: calculate redirection mark: ts=%#lx(now=%#lx),"
+		" att_no=%#x\n", mv->ts, jiffies, mv->att_no);
+
+	if ((r = crypto_shash_init(shash_desc)))
+		return r;
+	r = crypto_shash_update(shash_desc, (u8 *)&mv->att_no, sizeof(mv->att_no));
+	if (r)
+		return r;
+	return crypto_shash_finup(shash_desc, (u8 *)&mv->ts, sizeof(mv->ts),
+				  mv->hmac);
```

Comment:
It's better to use `crypto_shash_digest()` instead of the 3 calls.  Please update `__sticky_calc()` as well.

Reply:
Function `crypto_shash_finup()` is the short-hand itself (for `crypto_shash_update()` and `crypto_shash_final()`), so in fact we have 4 calls here (including two calls of `crypto_shash_update()` function - for two input values), but `crypto_shash_digest()` function is suitable for one input value only. The same situation in `__sticky_calc()`.